sigilrecorder.lic

```diff
@@ -0,0 +1,290 @@
+			  DRC.message("Adding #{roomid} to nogo due to lack of visible room number in LOOK.")
+			  add_to_nogo(roomid, @data)
+			  unstuck_me(roomid)
+			  next
+			end
+
+			#if the target room was reached successfully and has a roomid/image file (checked above) then start the 'perc sigil' process
+			if travel_result == true
+				sigils = find_sigils
+				record_sigil(sigils, @data) #write found sigils to file
+			else
+				#if the room was not reached successfully we assume it is unreachable and banlist it
+				DRC.message('Unable to reach target room. Adding to no-go list and moving to next room.')
+				add_to_nogo(roomid, @data) #write to nogo list in file
+			end #if
+		end #do
+
+		prune_final_list
+	end #initialize
+		
+	#returns the current season by parsing results of 'time' verb
+	def get_season
+		timestring = DRC.bput('time', /^It is currently.*/, /^The clerk says/)
+		if /^The clerk says/ =~ timestring #Annoyingly, this verb does something when near some NPCs. SIGH.
+		  DRC.message("The TIME verb does not work in this room.")
+		  add_to_nogo(Room.current.id, @data) #banlist annoying rooms where the 'time' verb works differently
+		  return false
+		end
+		/^It is currently (\w+) and it is.*$/ =~ timestring
+		return $1
+	end #get_season
+	
+	#detects the current region by comparing Room.current.image with region and city name patterns
+	def get_region
+		case Room.current.image.downcase
+		when /ilithi|Shard/i
+		  region = 'Shard'
+		when /zoluren|Crossing/i
+		  region = 'Crossing'
+		when /therengia|Riverhaven/i
+		  region = 'Riverhaven'
+		else
+		  DRC.message("Unsupported area. Sigilrecorder only supports areas where the associated map image file references Crossing, Shard, Riverhaven or their associated provinces. Current map image:#{Room.current.image}")
+		  exit
+		end #case
+		return region
+	end #get_region
+	
+	#builds a single array of all mapped sigil rooms for the current season then returns how many times roomid appears in that array
+	def check_room_data(roomid)
+		room_array = []
+		room_matches = 0
+		@sigil_list.each do |sigil|
+			@data.SigilInfo.dig(@region, sigil, @season).each do |number|
+				if number != nil
+					room_array.push(number)
+				end
+			end
+		end
+		
+		#remove nil values and then count the number of instances of roomid int the array
+		room_array.compact
+		room_matches = room_array.count(roomid)
+
+		DRC.message("Target room (#{roomid}) matches in data = #{room_matches}")
+		
+		#return the number of instances of roomid
+		return room_matches
+	end #check_room_data
+	
+	#attempts to navigate out of a room lacking important data
+	def unstuck_me(target_room)
+		#Check the exits to the room we think we're in in Mapdb. Store the exits in an array then loop through it.
+		exit_list = Map[target_room].wayto
+		exit_list.each do |rnum, dir|
+		  #flag to listen for room numbers appearing in 'peer' info
+		  Flags.add('sigil-room-detected', /Room Number: \d+/)
```

Comment:
🚧 The output `Room Number: #` is provided by another script, `roomnumbers`. Not everyone runs that script. Could you instead grab the room number from `Room.current.id`?

Reply:
Hm ok I am going to retool this a bit, then. I am going to ditch the flag and have it random-move and then check for Room.current.id != nil. I'll maybe cap it at like 3-5 attempts so I don't inadvertently get people into significant trouble.

This is primarily to get the user out of dead-end rooms that have mapping issues which are usually actual dead-ends but occasionally they're part of larger unmapped buildings/features, or areas with day/night issues. I thought about spamming mapping requests with *every* room I found with day-night issues but there's already so many sigil rooms there's no actual reason to bother other people so these rooms can be included. Even with multiple people sigil hunting there should be plenty of rooms to go around for all but the rarest sigils types in a region.

Reply:
I completely redid this function. I now store the 'start room' where the user began the script as a sort of anchor.

If the script ends a movement in a room without a roomid but did not "fail" to move to the intended room (as far as go2 is concerned) it then trusts that go2 got it to the desired room (since go2 threw no errors) but that the roomid is not visible for some reason. After adding the bad room to nogo, using this assumed roomid, I generate a path_to between the assumed roomid and the start room. Then, using wayto lookups and assumed room ids I slowly backtrace my steps toward the start room. At each stop along the way I check if the current room has a visible roomid, if so, I abort the loop, and the script continues to the next room in walk_list.

If the backtrace fails, the script will exit. Though I tested this in Under Wharfs in Riverhaven and it was able to avoid getting stuck down there so it's fairly robust (definitely not perfect).